### PR TITLE
Sensors/BMA2XX - New PR based on PR #697 changes

### PR DESCRIPTION
**Pull request by joshgrob**
@rymanluk @vrahane - I figured it was easier to make a new pull request with the complete changes and cleaner history.  The old PR is at https://github.com/apache/mynewt-core/pull/697, and I should have incorporated the comments.  Lastly, I kept in the change to the sensor.h file (adding <string.h>), because it was breaking the my test application even though sensor_test app works because of dependency differences.  I would like to get this change sooner rather than later.  Thank you for the time and help!

### CODE REVIEW

**Comment by rymanluk** [comment]
@joshgrob Could you please rebase it :) Basically what I do is
git fetch --all
git rebase -i origin/master (assuming that origin is git@github.com:apache/mynewt-core.git )

Now some editor will open and you will see list of patches which are going to be applied on top of origin/master. You should have there only two patches I guess. Remove patches which are not in your interest and save/close editor.

It should be it.

Then do force push on your remote branch (add option -f when pushing) and we can continue review.

**Comment by joshgrob** [comment]
rebase should be complete. Lemme know if something looks off, and thanks for the help in troubleshooting @rymanluk

**Comment by joshgrob** [comment]
@rymanluk Hope all is well.  Any additional thoughts on this PR?

File: hw/drivers/sensors/bma2xx/src/bma2xx.c [hunk 1]
```diff
@@ -0,0 +1,4940 @@
...
+    accel_data->new_data = raw_data[0] & 0x01;
+}
+
+static int get_accel_scale(enum bma2xx_model model,
+                            enum bma2xx_g_range g_range,
+                            float *accel_scale)
+{
+    switch (g_range) {
+        case BMA2XX_G_RANGE_2:
+            switch(model){
```
Comment (rymanluk) [inline hw/drivers/sensors/bma2xx/src/bma2xx.c]: maybe instead of switch(model) we could some const table of structs with custom parameters?

File: hw/sensor/creator/src/sensor_creator.c [hunk 2]
```diff
@@ -508,6 +543,51 @@ config_bma253_sensor(void)
...
+config_bma2xx_sensor(void)
+{
+    struct os_dev * dev;
+    struct bma2xx_cfg cfg;
+    int rc;
+
+    dev = os_dev_open("bma2xx_0", OS_TIMEOUT_NEVER, NULL);
+    assert(dev != NULL);
+
+    cfg.model = BMA2XX_BMA280;
```
Comment (rymanluk) [inline hw/sensor/creator/src/sensor_creator.c]: How I should use it if I have BMA253 or event both BMA280 and BMA253?

**Comment by sterlinghughes** [comment]
Hi @joshgrob -- would you be willing to respond to comments and rebase this pull request?  I'd like to merge it into Mynewt.

**Comment by joshgrob** [comment]
Hi @sterlinghughes many apologies for the delay in my comments.  Yes, I will try and respond by the end of the week or early next.

**Comment by sterlinghughes** [comment]
ping

File: hw/drivers/nimble/nrf52/src/ble_phy.c [hunk 3]
```diff
@@ -930,6 +930,7 @@ ble_phy_rx_start_isr(void)
     pdu_usecs = ble_phy_mode_pdu_start_off(ble_hdr->rxinfo.phy_mode) +
                 g_ble_phy_t_rxaddrdelay[ble_hdr->rxinfo.phy_mode];
     if (usecs < pdu_usecs) {
+        ticks--;
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/drivers/nimble/nrf52/src/ble_phy.c]: okay will try to do this.

File: hw/sensor/creator/src/sensor_creator.c [hunk 4]
```diff
@@ -184,10 +192,37 @@ static struct sensor_itf i2c_0_itf_lis = {
     .si_num  = 0,
     .si_addr = 0x18,
     .si_ints = {
-       { MYNEWT_VAL(BMA253_INT_PIN_HOST), MYNEWT_VAL(BMA253_INT_PIN_DEVICE),
-                                         MYNEWT_VAL(BMA253_INT_CFG_ACTIVE)},
-       { MYNEWT_VAL(BMA253_INT2_PIN_HOST), MYNEWT_VAL(BMA253_INT2_PIN_DEVICE),
-                                       MYNEWT_VAL(BMA253_INT_CFG_ACTIVE)}
+        { 26, MYNEWT_VAL(BMA2XX_INT_PIN_DEVICE),
```
[the remark this answers is not in the document]
Comment (joshgrob) [reply hw/sensor/creator/src/sensor_creator.c]: Okay, I will add this back.  When I reviewed with @vrahane earlier we had a reason to not make it configureable.  I can't remember now, but will change.

File: hw/drivers/sensors/bma2xx/src/bma2xx.c (hunk 1 above)
Comment (joshgrob) [reply hw/drivers/sensors/bma2xx/src/bma2xx.c]: Yeah, I was having second thoughts about this implementation, and reviewed with @vrahane and @wes3.  Given the limited number of models (7) we agreed it was okay to use this switch.  Is it okay if we keep this as is for now?

File: hw/sensor/creator/src/sensor_creator.c (hunk 2 above)
Comment (joshgrob) [reply hw/sensor/creator/src/sensor_creator.c]: There is two ways for this first implementation:
1.  Since Vipul's original BMA253 will persist alongside BMA2XX for a while the user can define both the BMA2XX_OFB and BMA253_OFB flags and two separate devices will be created.
2.  If the user wanted to just use the BMA2XX and create both a BMA280 and BMA253 sensor they would need to add a second device, bma2xx_cfg and sensor_itf in the appropriate BMA2XX_OFB sections.

Makes sense?

**Comment by joshgrob** [comment]
Okay I added my comments.  Will do the rebase after my comments are reviewed.  Will keep slack open and am reachable there for more detailed conversations.

File: hw/sensor/creator/src/sensor_creator.c (hunk 2 above)
Comment (rymanluk) [reply hw/sensor/creator/src/sensor_creator.c]: The issue I see here is that sensor_creator.c creates only one `bm2xx_0`, so if somebody want to make use of this driver for more then one sensor he needs to modify sensor_creator.c.  It would be nice to have possibility to configure it in other way. Anyway, this can be handled later.

File: hw/sensor/creator/src/sensor_creator.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (rymanluk) [reply hw/sensor/creator/src/sensor_creator.c]: I think this is related to my other comment on how to do proper configuration for bma2xx. I would put here TODO comment for this.

**Comment by sterlinghughes** [comment]
ping, can we do a new review or resolve?

**Comment by joshgrob** [comment]
Sorry, I know the delays here are lame.  I am traveling right now, and have it top of mind to test everything one last time before the final commit when I return and have my hardware setup. I return on Monday.  Many apologies, and thank you for the follow ups.

**Comment by joshgrob** [comment]
Hi @sterlinghughes @rymanluk, quick question.  I made the remaining changes, tested the code on my bmd300eval, and successfully did git pull --rebase origin master (from the apache mynewt-core repo).  All works well.  However, when I try to push to my fork it says it has diverged to much and I must to a git pull from that remote (fork/master in my repo).  I tried to a git pull --rebase fork master to align, but I am getting a lot of unexpected conflicts.

In short, what it is the best way to push my changes to pull request after I have successfully rebased my changes locally?

**Comment by ccollins476ad** [comment]
@joshgrob, the short answer is: you probably just need to do a `git push -f fork master` (i.e., a force push).

Here are some more details.  Since you have done a `git pull --rebase origin master`, then your local git history is incompatible with the remote history in your fork.  A normal (non-force) push can only append new commits on top of the remote history, it cannot interleave new commits.

Say your local and remote repos both have histories that look like this (`A#` are commits in apache-mynewt-core;`J#` are commits specific to this PR).

```
LOCAL       REMOTE
J3          J3
J2          J2
J1          J1
A4          A4
A3          A3
A2          A2
A1          A1
```

That is, your three commits are "on top" of the core repo.  Then you do a pull with rebase, which takes the latest from core and moves your commits on top of those:

```
LOCAL       REMOTE
J3          J3
J2          J2
J1          J1
A6          A4
A5          A3
A4          A2
A3          A1
A2
A1
```

Git can't reconcile these two histories.  Specifically `J1` has a different parent in your local repo than in the remote repo.  When you do a force push, you replace the remote history with an exact copy of your local history:
```
LOCAL       REMOTE
J3          J3
J2          J2
J1          J1
A6          A6
A5          A5
A4          A4
A3          A3
A2          A2
A1          A1
```

**Comment by joshgrob** [comment]
Thank for @ccollins476ad for the very helpful explanation.  If finally get it! Now, after I did a force to my fork the PR was automatically closed.  Is the expectation that I create a new PR base on the new commits in my fork?

**Comment by ccollins476ad** [comment]
@joshgrob - no, the PR should not have closed automatically!  It appears your PR-specific commits aren't in the remote branch anymore.  It seems you must have force-pushed a regular version of apache-mynewt-core which did not contain your commits.

Hopefully you still have your commits somewhere, perhaps in a separate directory?  If you do, you will want to push that onto your fork.  If you can't find your commits anywhere, then hopefully one of the reviewers has them locally.
